### ISSUE TITLE
FFXIII-2 fix branch

### ISSUE THREAD

**Issue by Sefie1999AD** [rebtd7/FF13Fix#1]
Thank you for your FFXIII fix! I would appreciate it if you could create a similar fix for FFXIII-2 as well. There is no Libra bug in FFXIII-2, but most of the same frame drop issues are still a problem. The framepacing is overly aggressive, but if you disable it (there is a fan-made framerate pacer remover available online), everything will happen in slow-motion if FPS falls between 40-59 (31-39 FPS will result in fast-motion). As in FFXIII, the 2D elements waste plenty of resources. Disabling minimap increases your FPS by about 10 frames, and in battles, frame drops happen because of the HUD, and large spell effects. The game also tends to crash on modern systems, but it can be made stable by disabling Steam cloud and using a 4 GB memory patch, which makes the game Large Address Aware.

If you do not have FFXIII-2 but you would be willing to expand your fix for FFXIII-2 as well, I would be glad to gift the game for you in Steam.

**Comment by Nucleoprotein** [comment]
FFXIII-2 problems are different from FFXIII so will be not easy way of making fork, maybe only frame pacer disable, but any FPS issues related to the way it renders frames are hard to be addressed.

**Comment by rebtd7** [comment]
Yeah, I'll probably release a first patch with only the controller scan removal, removed frame pacer and removed frame cap (if it can be done correctly without negative side effects like in FF13).

@Nucleoprotein any idea what can be done to further improve the performance here? What are the hard to be addressed issues?

**Comment by rebtd7** [comment]
I've uploaded an initial release for FF13-2. Please let me know if there are any issues.

The performance should be improved due to the removal of the frame pacer (hopefully without any negative side effect), but as the there is no vertex buffer fix here at the moment the effect isn't too drastic

**Comment by Sefie1999AD** [comment]
Thank you for the initial patch! While this release doesn't fix all the frame drops (on Ryzen 3700X, GTX 1070, 16 GB RAM), it removes the framerate pacer, and unlike the previous pacer remover from 2014, this one doesn't seem to cause slow-motion during frame drops.

The next step would probably be to track down why the mini-map, HUD and spell effects are so resource-heavy.

**Comment by rebtd7** [comment]
Turns out FF13-2 also keeps locking the same vertex buffer when drawing 2D elements to the screen, so the fix from OneTweakNG also applies to this game.

I've released a new version with it (1.3), it should considerably improve performance when 2D elements are being rendered to the screen.

The game still doesn't perform too well in some situations =/, but it should be much better than without any fix

**Comment by Nucleoprotein** [comment]
Yeah, in FF13 vertex buffer with length of 358400 bytes is causing slowdowns, it contains vertexes of all HUD elements like hand cursor etc. Switching this vertex to software processing ie. pointing it to system memory not GPU memory is resolving the issue, it seems constant uploads of this buffer to GPU are causing stalls. Strange the size is same, maybe they use same code in both games.

**Comment by Sefie1999AD** [comment]
The latest patch is fantastic. FFXIII-2 battles turned from constant fluctuation between 30-60 FPS to near-stable 60 FPS, and most of the time, not even Ruinga or other similar spells can drop the frame rate. Now that the minimap is no longer causing frame drops, towns stay at 60 FPS too, except for some parts at 4XX AF when the game is loading new areas. Excellent work! The game looks really beautiful and smooth at 1080p60fps with 4096x4096 shadows and 16x AA.

**Comment by Sefie1999AD** [comment]
I ran into a jerky camera issue at FFXIII-2's Chocobo races when the game runs at 60 FPS. Forcing the game to 30 FPS makes the camera smoother, but of course, the rest of the race or game aren't smoother. Can I contact you on Discord to send you a save game where you can access the Chocobo races?

**Comment by rebtd7** [comment]
Thanks for testing the fix

You can add me on discord (rebtd7#8605) but I'm not usually online
I think it is better if you attach the save here so if there's anyone interested try to fix the jerkiness

**Comment by Sefie1999AD** [comment]
Here's a save file that lets you access the Chocobo races. Instructions:

- Extract the dat file inside the zip archive into the folder C:\Users\[your user name]\AppData\Local\SquareEnix\FinalFantasyXIII-2\save (assuming you're not using Steam cloud for saving)
- The save file ff13_2-31.dat will only be detected if the previous file name on the listing (i.e. ff13_2-30.dat) exists. If it doesn't, you need to rename the dat file. For example, if your last file in the folder is called ff13_2-10.dat, rename this file into ff13_2-11.dat.
- When the save loads and you arrive at Historia Crux, enter Serendipity -Year Unknown-.
- Talk to the Chocobo girl at the counter. Choose Race, then choose any race, answer "Yes", and then select "Start Race". This will initiate the Chocobo race.
- Notice how the camera is terribly jerky unless you cap the FPS at 30.

[ff13_2-31.zip](https://github.com/rebtd7/FF13Fix/files/4377375/ff13_2-31.zip)

**Comment by rebtd7** [comment]
@Sefie1999AD Thanks for the save. 
This probably won't be an easy one to fix, unfortunately.

**Comment by chrcoluk** [comment]
if only rtss had a hotkey for framerate limiter, but alt-tabbing is now stable with this ff13fix, so the 30fps can be toggled manually in rtss when starting/stopping races.

**Comment by Sefie1999AD** [comment]
> if only rtss had a hotkey for framerate limiter, but alt-tabbing is now stable with this ff13fix, so the 30fps can be toggled manually in rtss when starting/stopping races.

@chrcoluk Bandicam has a hotkey that lets you toggle an FPS Limit, so you can limit FPS to e.g. 30 with a hotkey without having to alt-tab out of the game.

**Comment by Sefie1999AD** [comment]
@rebtd7 I found a bug with FFXIII-2 menus, such as the monster selection screen, equipment screen etc. With the FF13Fix, they are capped at 30 FPS for some reason.

**Comment by chrcoluk** [comment]
> > if only rtss had a hotkey for framerate limiter, but alt-tabbing is now stable with this ff13fix, so the 30fps can be toggled manually in rtss when starting/stopping races.
> 
> [@chrcoluk](https://github.com/chrcoluk) Bandicam has a hotkey that lets you toggle an FPS Limit, so you can limit FPS to e.g. 30 with a hotkey without having to alt-tab out of the game.

Nice to know as I dont routinely run rtss anymore, there is still definitely some sort of frame rate pacer loaded in the game though, as pointed out by sefie, there is still many areas capped at 30, and those areas usually have some instability as well, which heavily suggests a in game pacer.

I found a way to disable it completely, but the game has a ton of glitches that start occurring, which suggests a lot of game logic and in particular garbage collection is tied to frames.

With 60fps fully unlocked everywhere, libra screens, equip screen, monster screens etc.  I am guessing rebtd7 hit these issues.

Crystarium has a delay before it loads, it progressively gets longer the longer you run the game, and needs a full windows reboot to become (temporarily) quick again.  So restarting the game wont recover it, removing the pace disabler, makes it load instant again.
Garbage collection routines need much longer to complete, so is extra stuttering that doesnt occur with just ff13fix.  
In yashif area near entrance to village if you spawn a monster a loading spinning cd symbol appears and you are no longer able to move whilst it loads in the battle on my rtx 4080 super and 13700k, it needs about 5-6 seconds to load the battle, with the pacer enabled, its a micro stutter at worst.
Most areas take longer to load.
Loading monster images in the monster screen takes about 3-4x as long with the pacer disabled.

At a guess it looks like idle rendering frames are important for background processing.
